---
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Introduction

This is a very brief guide on how to extract other promoter sequences, e.g. of 
a different length relative to the TSS, from a genome of your choice. 

There are two steps here:
1. Getting the genome and genome annotation using `AnnotationHub`
2. Specify promoter lengths and extract

The packages I'm using are:

```{r load_r_packages, eval=TRUE, echo=TRUE, warning=FALSE}
suppressMessages(library(AnnotationHub))
suppressMessages(library(Biostrings))
suppressMessages(library(GenomicRanges))
suppressMessages(library(stringr))

options(stringsAsFactors = F)
```

## Getting the genome and genome annotation

Let's get custom promoter sequences for `mm10` using the `AnnotationHub` 
package. You can search for the right objects using the `AnnotationHub::query()`
function. 

```{r look_for_genome}
ah <- AnnotationHub::AnnotationHub()

# Let's look up a 2bit genome
query(ah, c("mm10", "2bit")) 
```

Let's get the genome:

```{r }
mm10 <- ah[["AH14005"]]

# Let's find the annotation too
query(ah, c("GRCm38", "gtf"))
mm10_gtf <- ah[["AH69533"]]

# Get unique genes in annotation
genes <- mm10_gtf[mm10_gtf$type == "gene" & 
                    seqnames(mm10_gtf) %in% c(1:19, "X", "Y")]
genes <- genes[!duplicated(genes$gene_name),]
```

For people needing `hg38`, I think the current `AnnotationHub` records are 
```
hg38 <- ah[["AH13963"]]
hg38_gtf <- ah[["AH79161"]]
```
but I recommend checking this with the `query(ah, c("GRCh38", "2bit"))` and 
`query(ah, c("GRCh38", "gtf"))` functions.

Next up: define how much sequence you want to extract. In this case, I'll go for
1 kb upstream and 0.5 kb downstream of the TSS. 

```{r extract_promoters}
# This defines how many bp upstream and downstream of TSSs you want to extract
upstream = 1000
downstream = 500
# This returns a GRanges object:
promoter_ranges <- GenomicRanges::promoters(genes, 
                                            upstream = upstream, 
                                            downstream = downstream)
# This extracts the sequences from the mm10 genome:
seqlevelsStyle(promoter_ranges) <- "UCSC"
promoter_seq <- Biostrings::getSeq(mm10, promoter_ranges)
# Naming these sequences is necessary for use in the data generation steps
names(promoter_seq) <- promoter_ranges$gene_name
```

Now I only want to keep the sequences that do not contain "N":

```{r only_keep_non_n}
promoter_seq <- promoter_seq[!stringr::str_detect(promoter_seq, "N")]
```

And now you save it with this:

```{r save_fasta, eval=FALSE, echo=TRUE}
Biostrings::writeXStringSet(promoter_seq, format = "fasta", 
                            filepath = "mm10_1500bp_promoters.fa")
```

I recommend converting the file to a `.fa.gz` file in the command line by 
running

```{bash gzip, eval=FALSE, echo=TRUE}
gzip mm10_1500bp_promoters.fa
```

And now you can use it in the data generation pipeline. 

## Used packages

```{r sessioninfo}
utils::sessionInfo()
```





