
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>How to run scover &#8212; scover  documentation</title>
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <script id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/language_data.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Choosing the right profile" href="profiles.html" />
    <link rel="prev" title="How to install scover prerequisites" href="how_to_install_scover_prerequisites.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <div class="section" id="how-to-run-scover">
<h1>How to run scover<a class="headerlink" href="#how-to-run-scover" title="Permalink to this headline">¶</a></h1>
<p>Usage:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>nextflow run <span class="o">[</span>options<span class="o">]</span> scover.nf <span class="o">[</span>arg...<span class="o">]</span>
</pre></div>
</div>
<p>Example command:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>nextflow run -profile lsf_gpu scover.nf <span class="se">\</span>
 --name test_run /
 --data data/mock_data.tsv <span class="se">\</span>
 --celldata data/mock_data_colData.tsv <span class="se">\</span>
 --tomtom resources/Mus_musculus.meme <span class="se">\</span>
 --num_calibrations <span class="m">30</span> <span class="se">\</span>
 --val_factor <span class="m">10</span> <span class="se">\</span>
 --epochs <span class="m">50</span> <span class="se">\</span>
 --num_errtest <span class="m">10</span> <span class="se">\</span>
 --motif_amount <span class="m">300</span> <span class="se">\</span>
 --motif_length <span class="m">12</span>
</pre></div>
</div>
<p><strong>scover</strong> requires a run name (<code class="code docutils literal notranslate"><span class="pre">--name</span></code>), a pooled dataset (<code class="code docutils literal notranslate"><span class="pre">--data</span></code>), and pool cell type annotations (<code class="code docutils literal notranslate"><span class="pre">--celldata</span></code>).
For creating a pooled dataset, find the right guide in the <a class="reference external" href="https://scover.readthedocs.io/en/latest/index.html">documentation</a>.
Another important argument is <code class="code docutils literal notranslate"><span class="pre">--tomtom</span></code>, for specifying
the motif <code class="code docutils literal notranslate"><span class="pre">.meme</span></code> database file to align the found motifs to. I have included <code class="code docutils literal notranslate"><span class="pre">Mus_musculus.meme</span></code> and <code class="code docutils literal notranslate"><span class="pre">Homo_sapiens.meme</span></code>
from CIS-BP[1] in the <code class="code docutils literal notranslate"><span class="pre">resources</span></code> directory.</p>
<p>Before running <strong>scover</strong> you will probably want to identify the best <code class="code docutils literal notranslate"><span class="pre">-profile</span></code> to run with. This will define the executor
used by <strong>scover</strong>. <a class="reference external" href="profiles.html">This page</a> has more information on the profiles specified for <strong>scover</strong>. <a class="reference external" href="https://scover.readthedocs.io/en/latest/profiles.html">See this page</a> for options and customisation.</p>
<p>Minimal command:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>nextflow run scover.nf <span class="se">\</span>
 --name test_run <span class="se">\</span>
 --data data/mock_data.tsv <span class="se">\</span>
 --celldata data/mock_data_colData.tsv <span class="se">\</span>
 --tomtom resources/Mus_musculus.meme
</pre></div>
</div>
<p>Important:
It might seem like it is stuck during training as the line with</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="o">[</span><span class="m">62</span>/3eb656<span class="o">]</span> process &gt; scover             <span class="o">[</span>  <span class="m">0</span>%<span class="o">]</span> <span class="m">0</span> of <span class="m">1</span>
</pre></div>
</div>
<p>keeps saying <code class="code docutils literal notranslate"><span class="pre">[</span>&#160; <span class="pre">0%]</span></code>. This happens because the training stage only contains one
neural network step, and it will switch to <code class="code docutils literal notranslate"><span class="pre">[100%]</span></code> once the training is complete.
There is a way to find out how far into training the network is, but it is not entirely straightforward:
Open a second terminal window and <code class="code docutils literal notranslate"><span class="pre">cd</span></code> into the <code class="code docutils literal notranslate"><span class="pre">scover/work</span></code> directory. From there, you
can <code class="code docutils literal notranslate"><span class="pre">cd</span></code> into the directory starting with the code shown in front of the specific
task (in this case, <code class="code docutils literal notranslate"><span class="pre">62/3eb656</span></code>, so the directory will be something like <code class="code docutils literal notranslate"><span class="pre">scover/work/62/3eb656aae5e84c420b7aa267dfeb57</span></code>).
Then, by running <code class="code docutils literal notranslate"><span class="pre">tail</span> <span class="pre">.command.log</span></code> you can get an idea of how far along the training is.</p>
<div class="section" id="run-an-example-dataset">
<h2>Run an example dataset<a class="headerlink" href="#run-an-example-dataset" title="Permalink to this headline">¶</a></h2>
<p>Try to run <strong>scover</strong> with an example dataset by <a class="reference external" href="profiles.html">choosing your profile to run with</a>, making sure
you are in an environment that can run both Nextflow and Singularity, and then run the following from the <code class="code docutils literal notranslate"><span class="pre">scover</span></code> directory:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>nextflow run -profile local_gpu scover.nf <span class="se">\</span>
 --data data/small_dataset.tsv <span class="se">\</span>
 --celldata data/small_dataset_colData.tsv <span class="se">\</span>
 --name example_run <span class="se">\</span>
 --epochs <span class="m">8</span> <span class="se">\</span>
 --num_candidates <span class="m">8</span> <span class="se">\</span>
 --num_calibrations <span class="m">4</span> <span class="se">\</span>
 --tomtom resources/Mus_musculus.meme
</pre></div>
</div>
<p>Note that you should change <code class="code docutils literal notranslate"><span class="pre">local_gpu</span></code> to whatever profile you will use. This should create a folder in the <code class="code docutils literal notranslate"><span class="pre">scanem/output</span></code> directory named <code class="code docutils literal notranslate"><span class="pre">example_run</span></code> with all the output files. Note that most of the results will not make sense here, this is a dataset that contains the right data structure but not enough data for the network to converge properly.</p>
</div>
<div class="section" id="arguments">
<h2>Arguments<a class="headerlink" href="#arguments" title="Permalink to this headline">¶</a></h2>
<p><code class="code docutils literal notranslate"><span class="pre">--name</span></code>
The name of the current experiment.
Default value: <code class="code docutils literal notranslate"><span class="pre">experiment</span></code></p>
<p><code class="code docutils literal notranslate"><span class="pre">--data</span></code>
The path to the dataset (relative to <code class="code docutils literal notranslate"><span class="pre">scover.nf</span></code>) to train the network. Add your data in the example tab-separated format into the <cite>/data</cite> folder.</p>
<p><code class="code docutils literal notranslate"><span class="pre">--celldata</span></code>
The path to the dataset annotation (relative to <code class="code docutils literal notranslate"><span class="pre">scover.nf</span></code>) to annotate the cells.
Add your data in the example format (see default file) into the <code class="code docutils literal notranslate"><span class="pre">/data</span></code> folder.
For more information, see the data generation guides in the documentation.</p>
<p><code class="code docutils literal notranslate"><span class="pre">--tomtom</span></code>
The path to the <code class="code docutils literal notranslate"><span class="pre">.meme</span></code> format motif database to align found motifs to.
Default value: <code class="code docutils literal notranslate"><span class="pre">resources/Mus_musculus.meme</span></code>.</p>
<p><code class="code docutils literal notranslate"><span class="pre">--num_calibrations</span></code>
The number of randomly intialized calibrations for hyperparameter optimization.
Default value: <code class="code docutils literal notranslate"><span class="pre">30</span></code></p>
<p><code class="code docutils literal notranslate"><span class="pre">--num_candidates</span></code>
The number of candidate models with optimal initial parameters that should be run.
Default value: <code class="code docutils literal notranslate"><span class="pre">10</span></code></p>
<p><code class="code docutils literal notranslate"><span class="pre">--val_factor</span></code>
Sets K for the K-fold cross-validation.
Default value: <code class="code docutils literal notranslate"><span class="pre">10</span></code>.</p>
<p><code class="code docutils literal notranslate"><span class="pre">--epochs</span></code>
Sets the number of epochs (1 epoch = 1 forward and backward cycle through the entire training set) the network goes through.
Default value: <code class="code docutils literal notranslate"><span class="pre">100</span></code></p>
<p><code class="code docutils literal notranslate"><span class="pre">--batch_size</span></code>
The size of one training batch (amount of sequences and the corresponding outputs). After one batch, the network will update its parameters through back-propagation.
Default value: <code class="code docutils literal notranslate"><span class="pre">32</span></code></p>
<p><code class="code docutils literal notranslate"><span class="pre">--motif_length</span></code>
The length of each individual regulatory motif.
Default value: <code class="code docutils literal notranslate"><span class="pre">12</span></code></p>
<p><code class="code docutils literal notranslate"><span class="pre">--motif_amount</span></code>
The amount of regulatory motifs to look for.
Default value: <code class="code docutils literal notranslate"><span class="pre">300</span></code></p>
<p><code class="code docutils literal notranslate"><span class="pre">--sigma_motifs_min</span></code>
sigma_motifs will be drawn from loguniform(sigma_motifs_min, sigma_motifs_max).
Subsequently, sigma_motifs acts as the standard deviation of the normal
distribution from which convolutional kernel coefficients are drawn.
Default value: <code class="code docutils literal notranslate"><span class="pre">1e-7</span></code></p>
<p><code class="code docutils literal notranslate"><span class="pre">--sigma_motifs_max</span></code>
sigma_motifs will be drawn from loguniform(sigma_motifs_min, sigma_motifs_max).
Subsequently, sigma_motifs acts as the standard deviation of the normal
distribution from which convolutional kernel coefficients are drawn.
Default value: <code class="code docutils literal notranslate"><span class="pre">1e-3</span></code></p>
<p><code class="code docutils literal notranslate"><span class="pre">--sigma_net_min</span></code>
sigma_net will be drawn from loguniform(sigma_net_min, sigma_net_max).
Subsequently, sigma_net acts as the standard deviation of the normal
distribution from which neural network layer coefficients are drawn.
Default value: <code class="code docutils literal notranslate"><span class="pre">1e-5</span></code></p>
<p><code class="code docutils literal notranslate"><span class="pre">--sigma_net_max</span></code>
sigma_net will be drawn from loguniform(sigma_net_min, sigma_net_max).
Subsequently, sigma_net acts as the standard deviation of the normal
distribution from which neural network layer coefficients are drawn.
Default value: <code class="code docutils literal notranslate"><span class="pre">1e-2</span></code></p>
<p><code class="code docutils literal notranslate"><span class="pre">--epsilon_min</span></code>
The learning rate will be drawn from loguniform(epsilon_min, epsilon_max).
Default value: <code class="code docutils literal notranslate"><span class="pre">5e-4</span></code></p>
<p><code class="code docutils literal notranslate"><span class="pre">--epsilon_max</span></code>
The learning rate will be drawn from loguniform(epsilon_min, epsilon_max).
Default value: <code class="code docutils literal notranslate"><span class="pre">5e-2</span></code></p>
<p><code class="code docutils literal notranslate"><span class="pre">--opt</span></code>
The pytorch optimizer to use. Options include <code class="code docutils literal notranslate"><span class="pre">SGD</span></code> and <code class="code docutils literal notranslate"><span class="pre">Adam</span></code>.
Default value: <code class="code docutils literal notranslate"><span class="pre">SGD</span></code></p>
<p><strong>References</strong></p>
<p>[1] Weirauch, M. T., Yang, A., Albu, M., Cote, A. G., Montenegro-Montero, A., Drewe, P., … &amp; Zheng, H. (2014). Determination and inference of eukaryotic transcription factor sequence specificity. Cell, 158(6), 1431-1443.</p>
</div>
</div>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<p class="logo">
  <a href="index.html">
    <img class="logo" src="_static/scover_logo.png" alt="Logo"/>
    
  </a>
</p>








<h3>Navigation</h3>
<p class="caption"><span class="caption-text">Dataset creation:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="how_to_prepare_scover_input_using_scRNA_seq.html">How to generate scover input using scRNA-seq data</a></li>
<li class="toctree-l1"><a class="reference internal" href="how_to_prepare_scover_input_using_scATAC_seq.html">How to generate scover input using scATAC-seq data</a></li>
<li class="toctree-l1"><a class="reference internal" href="how_to_prepare_scover_input_using_create_dataset.html">How to generate scover input using scRNA-seq data and create_dataset.R</a></li>
<li class="toctree-l1"><a class="reference internal" href="how_to_extract_promoter_sequences.html">How to extract promoter sequences</a></li>
</ul>
<p class="caption"><span class="caption-text">Running scover:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="how_to_install_scover_prerequisites.html">How to install scover prerequisites</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">How to run scover</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#run-an-example-dataset">Run an example dataset</a></li>
<li class="toctree-l2"><a class="reference internal" href="#arguments">Arguments</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="profiles.html">Profiles</a></li>
</ul>
<p class="caption"><span class="caption-text">Analysing scover output:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="how_to_analyse_scover_output.html">How to analyse scover output</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
      <li>Previous: <a href="how_to_install_scover_prerequisites.html" title="previous chapter">How to install scover prerequisites</a></li>
      <li>Next: <a href="profiles.html" title="next chapter">Choosing the right profile</a></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2020, Jacob Hepkema.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 3.3.1</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="_sources/how_to_run_scover.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>