
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>scanem &#8212; scanem  documentation</title>
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/language_data.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="How to prepare scanem input using scRNA-seq data" href="how_to_prepare_scanem_input_using_scRNA_seq.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <div class="section" id="scanem">
<h1>scanem<a class="headerlink" href="#scanem" title="Permalink to this headline">¶</a></h1>
<p>Welcome to the scanem documentation.</p>
<p><strong>scanem</strong> is a convolutional neural network (CNN) for <em>de novo</em> inference of cis-regulatory motifs from single-cell data.
It finds weights for these motifs across pseudo-bulks and also reports the ‘impact’ of each motif.
The network is written in pytorch, with the downstream analyses written in R.
Running the network and running the downstream analysis is implemented in a Nextflow pipeline.
Furthermore, motifs are aligned with Tomtom from the MEME suite[1].
scanem requires that cells are annotated for cell type (or other category).
For scRNA-seq data TSS annotation is required, as the promoter sequences are obtained directly from genomic sequence relative to the TSS.</p>
<div class="toctree-wrapper compound">
<p class="caption"><span class="caption-text">Contents:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="how_to_prepare_scanem_input_using_scRNA_seq.html">How to generate scanem input using scRNA-seq data</a></li>
<li class="toctree-l1"><a class="reference internal" href="how_to_prepare_scanem_input_using_create_dataset.html">How to generate scanem input using scRNA-seq data and create_dataset.R</a></li>
<li class="toctree-l1"><a class="reference internal" href="how_to_analyse_scanem_output.html">How to analyse scanem output</a></li>
</ul>
</div>
<div class="section" id="references">
<h2>References<a class="headerlink" href="#references" title="Permalink to this headline">¶</a></h2>
<p>[1]  Gupta, S., Stamatoyannopoulos, J. A., Bailey, T. L., &amp; Noble, W. S. (2007). Quantifying similarity between motifs. Genome biology, 8(2), R24.</p>
</div>
</div>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<p class="logo">
  <a href="#">
    <img class="logo" src="_static/scanem_logo.png" alt="Logo"/>
    
  </a>
</p>








<h3>Navigation</h3>
<p class="caption"><span class="caption-text">Contents:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="how_to_prepare_scanem_input_using_scRNA_seq.html">How to generate scanem input using scRNA-seq data</a></li>
<li class="toctree-l1"><a class="reference internal" href="how_to_prepare_scanem_input_using_create_dataset.html">How to generate scanem input using scRNA-seq data and create_dataset.R</a></li>
<li class="toctree-l1"><a class="reference internal" href="how_to_analyse_scanem_output.html">How to analyse scanem output</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="#">Documentation overview</a><ul>
      <li>Next: <a href="how_to_prepare_scanem_input_using_scRNA_seq.html" title="next chapter">How to prepare scanem input using scRNA-seq data</a></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2020, Jacob Hepkema.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 3.2.1</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="_sources/index.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>